Proposal: add --add-registry and --block-registry options to docker daemon

miminar opened the pull request:
The first option allows to prepend additional registries to a public one. These will be queried one after another in the order given when pulling images.

For example:

```
--add-registry=foo.com
--add-registry=bar.com
```

Would cause a

```
docker pull myapp
```

to search

```
foo.com/myapp
bar.com/myapp
docker.io/myapp
```

The second option builds a list of banned registries. Docker daemon will prevent user from any communication with such registries. This option recognizes a special keywords "public" denoting public Docker registry and "all" causing all registries except for those added with `--add-registry` flag to be blocked.

So for example adding

```
--block-registry=public
```

to flags of previous example would result in searching the same registries except for `docker.io/myapp`.

This would allow companies and vendors to specify registries which contain content that the companies will not allowed to be stored at docker.io.

Registry mirrors specified with `--registry-mirror` apply just to official repository.

This is an attempt to resurrect #5821

All the images are now prefixed with their source registry:

```
 $ docker images
REPOSITORY                                             TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker.io/ubuntu                                       14.04               5ba9dab47459        43 hours ago        188.3 MB
docker.io/ubuntu                                       14.04.1             5ba9dab47459        43 hours ago        188.3 MB
docker.io/ubuntu                                       latest              5ba9dab47459        43 hours ago        188.3 MB
docker.io/ubuntu                                       trusty              5ba9dab47459        43 hours ago        188.3 MB
docker-registry.usersys.redhat.com/vpavlin/rheltools   latest              94bcddfdb384        10 days ago         450.4 MB
registry.access.redhat.com/rhel7                       0                   bef54b8f8a2f        7 months ago        139.6 MB
registry.access.redhat.com/rhel7                       0-23                bef54b8f8a2f        7 months ago        139.6 MB
registry.access.redhat.com/rhel7                       latest              bef54b8f8a2f        7 months ago        139.6 MB

$ docker search ubuntu    # last 2 columns omitted
NAME                                    DESCRIPTION                                     STAR
docker.io/ubuntu                        Official Ubuntu base image                      1227
docker.io/dockerfile/ubuntu             Trusted automated Ubuntu (http://www.ubunt...   40
docker.io/tutum/ubuntu                  Ubuntu image with SSH access. For the root...   14
docker.io/alsanium/ubuntu               Ubuntu Core image for Docker                    1
docker.io/sylvainlasnier/ubuntu         Ubuntu 14.10 root docker images with commo...   1
docker.io/tleyden5iwx/ubuntu-cuda       Ubuntu 14.04 with CUDA drivers pre-installed    1
docker.io/mhutter/ubuntu                Base Ubuntu Image                               0
docker.io/partlab/ubuntu                Simple Ubuntu docker images.                    0
registry.my.org/custom/ubuntu                                                           0
registry.my.org/my/ubuntu  
```

Commit:
Add --add-registry and --block-registry options to docker daemon

    The first option allows to prepend additional registries to a public
    one. These will be queried one after another - in the same order
    they were given - when pulling images.

    For example:

        --add-registry=foo.com
        --add-registry=bar.com

    Would cause a

        docker pull myapp

    to search

        foo.com/myapp
        bar.com/myapp
        docker.io/myapp

    The second option builds a list of banned registries. Docker daemon
    will prevent user from any communication with such registries. This
    option recognizes a special keyword "public" denoting public Docker
    registry.

    So for example adding

        --block-registry=public

    to flags of previous example would result in searching the same
    registries except for `docker.io/myapp`

    Both flags are marked as experimental in man pages.

    This would allow companies and vendors to specify registries which
    contain content that the companies will not allowed to be stored at
    docker.io.

    The FROM statement in Dockerfile keeps its old behaviour. If the
    repository is missing registry, it will be pulled from the official
    one.

Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
Known shortcomings:
 * missing tests
 * search command does not show origin

SvenDowideit commented on docs/man/docker.1.md:
```diff
@@ -34,6 +34,9 @@ unix://[/path/to/socket] to use.
    tcp://host:port, unix:///path/to/socket, fd://* or fd://socketfd.
 
+**--add-registry**=[]
+  Each given registry will be prepended to a list of registries queried during image pulls or searches. The last registry given will be queried first. They will be treated as insecure. Registry mirrors won't apply to them.
```
please line wrap docs at about 80 chars

SvenDowideit replied:
I think that the `--registry-mirror` option queried in the order given - if that's the case, whats the reason this option starts at the last registry given?

Should we really default to insecure registry?

miminar replied:
LIFO policy in this case seems more natural to me. If you consider the list of queried registries as a list with one initial value `docker.io`, then each addition inserts registry on the first position. FIFO policy, on the other hand, inserts to a last but one position - before `docker.io`, which is kind of awkward to imagine and explain. But I agree it's confusing that these similar flags behave differently. I guess nothing can be done about `--registry-mirror` now. So I'll make `--add-registry` behave the same.

miminar replied:
The support for secured registries can be added later since most of private registries don't need to be protected. I'll look into it.

miminar commented:
The failed build ends with

    http: TLS handshake error from 127.0.0.1:37914: tls: client didn't provide a certificate
signal: broken pipe

which is kind of weird and most probably unrelated to the patch. Can it be somehow rescheduled?

Commit:
Images from public repository fully qualified

Local name of images from public repository is now prefixed with
docker.io.

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Display search results with index name of their origin

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Sort search result by index name

Images found by search command are now sorted by:
    1. index name of their repository
    2. number of stars (if they belong to public Docker registry)
    3. repository name

Example (last two columns omitted for brewity):
    $ docker search ubuntu
    NAME                                    DESCRIPTION                                     STAR
    docker.io/ubuntu                        Official Ubuntu base image                      1227
    docker.io/dockerfile/ubuntu             Trusted automated Ubuntu (http://www.ubunt...   40
    docker.io/tutum/ubuntu                  Ubuntu image with SSH access. For the root...   14
    docker.io/alsanium/ubuntu               Ubuntu Core image for Docker                    1
    docker.io/sylvainlasnier/ubuntu         Ubuntu 14.10 root docker images with commo...   1
    docker.io/tleyden5iwx/ubuntu-cuda       Ubuntu 14.04 with CUDA drivers pre-installed    1
    docker.io/mhutter/ubuntu                Base Ubuntu Image                               0
    docker.io/partlab/ubuntu                Simple Ubuntu docker images.                    0
    registry.my.org/custom/ubuntu                                                           0
    registry.my.org/my/ubuntu                                                               0

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
Why Jenkins need to run tests for `docker-py` which does not belong to `docker` repository? IMHO they should be run against [docker-py](https://github.com/docker/docker-py) and master branch of this repository.

Commit:
Support "all" keyword with --block-registry flag

Supplying `--block-registry=all` to docker daemon will prevent docker
from contacting any registry except for those added with
`--add-registry` flag.

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
/ping @dmcgowan @dmp42 @icecrime @crosbymichael @shykes
Do you see this as a viable solution to *private registries problem*? Or shall this follow #8329 more closely? I'm open to critic and proposals for improvement. I'd just like to see some working solution in the next Docker version.

icecrime commented:
Mmm I thought this had been discussed with @rhatdan and that we agreed not to merge this upstream. I am missing something?

rhatdan commented:
Well we want to show what we have done, and would expect behaviour to work like this in V2.

Commit:
Added tests for additional registries

It tests behaviour of pull, push and build operations with
`--add-registry` and `--block-registry` flags passed to docker daemon.

Signed-off-by: Michal Minar <miminar@redhat.com>

thaJeztah commented:
Similar proposal; https://github.com/docker/docker/issues/8326 (and related https://github.com/docker/docker/issues/8329)

rhatdan commented:
I think it is much simpler to allow admins to configure which registries they allow/disallow then trying to change the behaviour of pull/push/search.

ByteFlinger commented:
It would be nice to further specify which type of function (push, pull) you want to add/block registries.
I don't see a reason in many enterprise scenarios to keep a copy of an image which already exists in docker hub in a private registry yet you still might want to prevent pushes to a public registry. In such a scenario one would want to be able to pull from a public repo but not push to it.

rhatdan commented:
@ByteFlinger Yes that seems reasonable.

dmp42 commented:
@rhatdan @miminar there is a lot of namespace & trust work going on right now (involving people from RH - Andy & Vincent), which I feel has a much better chance of being merged than this (which always was flagged as "strawman/not-going-to-be-merged").
Also, the feeling now is that new ideas impacting registries should first go through design with the distribution team (and yes, that works great: #11109)

Thus I would like to close this one for bookkeeping.

If you think this is really wrong, please state so and I will re-open.

Hope that helps.

rhatdan commented:
That is fine.  As long as the "CLI Interface" gets strong consideration, since we are now shipping it.

finkr commented:
This feature was ultimately refused ([PR #10411](https://github.com/moby/moby/pull/10411/)).

The [rational](https://github.com/moby/moby/issues/11816#issuecomment-86732779) was:
> Like pointed out earlier (#11815), this would fragment the namespace, and hurt the community pretty badly, making dockerfiles no longer portable.

mbentley commented:
I found that rationale to be poorly founded.  Changing the default registry when you leave off the FQDN, where Docker currently defaults to Docker Hub, would "[make] dockerfiles no longer portable".  The ability to control which registries you are able to push/pull to/from would not do so.  It would be a level of access control to prevent unwanted/untrusted registries from being accessed.

joshuafc commented:
Any one knowns whether there a forked repo support the feature?

joshuafc commented:
@mbentley As a user, I known : 1.Any registries at Internet are unwanted.  2.Only the registry I created  at LAN is trusted and wanted.

mephmanx commented:
joshuafc

Did you ever find anything that worked like this?
